File: src/parser.ts
/* Copyright (c) 2013-2022 Richard Rodger, MIT License */

/*  parser.ts
 *  Parser implementation, converts the lexer tokens into parsed data.
 */

import type {
  Config,
  Context,
  Options,
  ParsePrepare,
  Parser,
  Rule,
  RuleDefiner,
  RuleSpec,
  RuleSpecMap,
} from './types'

import { EMPTY } from './types'

import {
  JsonicError,
  S,
  badlex,
  deep,
  filterRules,
  keys,
  srcfmt,
  tokenize,
  // log_stack,
  values,
} from './utility'

import { makeNoToken, makeLex, makePoint, makeToken } from './lexer'

import { makeRule, makeNoRule, makeRuleSpec } from './rules'

class ParserImpl implements Parser {
  options: Options
  cfg: Config
  rsm: RuleSpecMap = {}

  constructor(options: Options, cfg: Config) {
    this.options = options
    this.cfg = cfg
  }

  // TODO: ensure chains properly, both for create and extend rule
  // Multi-functional get/set for rules.
  rule(
    name?: string,
    define?: RuleDefiner | null
  ): RuleSpec | RuleSpecMap | undefined {
    // If no name, get all the rules.
    if (null == name) {
      return this.rsm
    }

    // Else get a rule by name.
    let rs: RuleSpec = this.rsm[name]

    // Else delete a specific rule by name.
    if (null === define) {
      delete this.rsm[name]
    }

    // Else add or redefine a rule by name.
    else if (undefined !== define) {
      rs = this.rsm[name] = this.rsm[name] || makeRuleSpec(this.cfg, {})
      rs = this.rsm[name] = define(this.rsm[name], this) || this.rsm[name]
      rs.name = name

      // Ensures jsonic.rule can chain
      return undefined
    }

    return rs
  }

  start(src: string, jsonic: any, meta?: any, parent_ctx?: any): any {
    let root: Rule

    let endtkn = makeToken(
      '#ZZ',
      tokenize('#ZZ', this.cfg),
      undefined,
      EMPTY,
      makePoint(-1)
    )

    let notoken = makeNoToken()

    let ctx: Context = {
      uI: 0,
      opts: this.options,
      cfg: this.cfg,
      meta: meta || {},
      src: () => src, // Avoid printing src
      root: () => root,
      plgn: () => jsonic.internal().plugins,
      inst: () => jsonic,
      rule: {} as Rule,
      sub: jsonic.internal().sub,
      xs: -1,
      v2: endtkn,
      v1: endtkn,
      t0: notoken,
      t1: notoken,
      tC: -2, // Prepare count for 2-token lookahead.
      kI: -1,
      rs: [],
      rsI: 0,
      rsm: this.rsm,
      log: undefined,
      F: srcfmt(this.cfg),
      use: {},
      NOTOKEN: notoken,
      NORULE: {} as Rule,
    }

    ctx = deep(ctx, parent_ctx)

    let norule = makeNoRule(ctx)
    ctx.NORULE = norule
    ctx.rule = norule

    // makelog(ctx, meta)
    if (meta && S.function === typeof meta.log) {
      ctx.log = meta.log
    }

    this.cfg.parse.prepare.forEach((prep: ParsePrepare) =>
      prep(jsonic, ctx, meta)
    )

    // Special case - avoids extra per-token tests in main parser rules.
    if ('' === src) {
      if (this.cfg.lex.empty) {
        return this.cfg.lex.emptyResult
      } else {
        throw new JsonicError(S.unexpected, { src }, ctx.t0, norule, ctx)
      }
    }

    let lex = badlex(makeLex(ctx), tokenize('#BD', this.cfg), ctx)
    let startspec = this.rsm[this.cfg.rule.start]

    if (null == startspec) {
      return undefined
    }

    let rule = makeRule(startspec, ctx)

    root = rule

    // Maximum rule iterations (prevents infinite loops). Allow for
    // rule open and close, and for each rule on each char to be
    // virtual (like map, list), and double for safety margin (allows
    // lots of backtracking), and apply a multipler option as a get-out-of-jail.
    let maxr =
      2 * keys(this.rsm).length * lex.src.length * 2 * ctx.cfg.rule.maxmul

    // Process rules on tokens
    let kI = 0

    // This loop is the heart of the engine. Keep processing rule
    // occurrences until there's none left.
    while (norule !== rule && kI < maxr) {
      ctx.kI = kI
      ctx.rule = rule

      ctx.log && ctx.log('', ctx.kI + ':')

      if (ctx.sub.rule) {
        ctx.sub.rule.map((sub) => sub(rule, ctx))
      }

      rule = rule.process(ctx, lex)

      // ctx.log && log_stack(ctx, rule, lex)
      ctx.log && ctx.log(S.stack, ctx, rule, lex)

      kI++
    }

    // TODO: option to allow trailing content
    if (endtkn.tin !== lex.next(rule).tin) {
      throw new JsonicError(S.unexpected, {}, ctx.t0, norule, ctx)
    }

    // NOTE: by returning root, we get implicit closing of maps and lists.
    const result = ctx.root().node

    if (this.cfg.result.fail.includes(result)) {
      throw new JsonicError(S.unexpected, {}, ctx.t0, norule, ctx)
    }

    return result
  }

  clone(options: Options, config: Config) {
    let parser = new ParserImpl(options, config)

    // Inherit rules from parent, filtered by config.rule
    parser.rsm = Object.keys(this.rsm).reduce(
      (a, rn) => ((a[rn] = filterRules(this.rsm[rn], this.cfg)), a),
      {} as any
    )

    parser.norm()

    return parser
  }

  norm() {
    values(this.rsm).map((rs: RuleSpec) => rs.norm())
  }
}

const makeParser = (...params: ConstructorParameters<typeof ParserImpl>) =>
  new ParserImpl(...params)

export { makeRule, makeRuleSpec, makeParser }

File: go/parser.go
package jsonic

import (
	"math"
	"strconv"
	"strings"
)

// Context holds the parse state, matching the TypeScript Context type.
type Context struct {
	UI       int               // Unique rule ID counter (TS: uI)
	T0       *Token            // Current token (TS: t0)
	T1       *Token            // Next token / lookahead (TS: t1)
	V1       *Token            // Previous token (TS: v1)
	V2       *Token            // Previous previous token (TS: v2)
	RS       []*Rule           // Rule stack (TS: rs)
	RSI      int               // Rule stack index (TS: rsI)
	RSM      map[string]*RuleSpec // Rule spec map (TS: rsm)
	KI       int               // Iteration counter (TS: kI)
	Rule     *Rule             // Current parsing rule (TS: rule)
	Meta     map[string]any    // Parse metadata (TS: meta)
	LexSubs  []LexSub          // Lex event subscribers (TS: sub.lex)
	RuleSubs []RuleSub         // Rule event subscribers (TS: sub.rule)
	ParseErr *Token            // Error token, halts parse

	// Fields matching TS Context:
	Opts     *Options          // Jsonic instance options (TS: opts)
	Cfg      *LexConfig        // Jsonic instance config (TS: cfg)
	Src      string            // Source text being parsed (TS: src)
	Inst     *Jsonic           // Current Jsonic instance (TS: inst)
	U        map[string]any    // Custom plugin data bag (TS: u)
	Root     *Rule             // Root rule (TS: root)
	TC       int               // Token count (TS: tC)
	F        func(any) string  // Format value as string (TS: F)
	Log      func(...any)      // Debug logger (TS: log)
	NOTOKEN  *Token            // Sentinel no-token (TS: NOTOKEN)
	NORULE   *Rule             // Sentinel no-rule (TS: NORULE)
}

// Parser orchestrates the parsing process.
type Parser struct {
	Config        *LexConfig
	RSM           map[string]*RuleSpec
	MaxMul        int               // Max rule occurrence multiplier. Default: 3.
	ErrorMessages map[string]string  // Custom error message templates.
	Hints         map[string]string  // Explanatory hints per error code.
	ErrTag        string             // Custom error tag (TS: errmsg.name). Default: "jsonic".
}

// NewParser creates a parser with default configuration.
func NewParser() *Parser {
	cfg := DefaultLexConfig()
	rsm := make(map[string]*RuleSpec)
	buildGrammar(rsm, cfg)
	// Copy global error messages as defaults.
	msgs := make(map[string]string, len(errorMessages))
	for k, v := range errorMessages {
		msgs[k] = v
	}
	return &Parser{Config: cfg, RSM: rsm, MaxMul: 3, ErrorMessages: msgs}
}

// Start parses the source string and returns the result.
// Returns a *JsonicError if parsing fails.
func (p *Parser) Start(src string) (any, error) {
	return p.startParse(src, nil, nil, nil, nil)
}

// StartMeta parses the source string with metadata, subscriptions, and
// an optional Jsonic instance reference (for Context.Inst).
func (p *Parser) StartMeta(src string, meta map[string]any, lexSubs []LexSub, ruleSubs []RuleSub) (any, error) {
	return p.startParse(src, meta, lexSubs, ruleSubs, nil)
}

// startParse is the internal entry point that populates the full Context.
func (p *Parser) startParse(src string, meta map[string]any, lexSubs []LexSub, ruleSubs []RuleSub, inst *Jsonic) (any, error) {
	if src == "" {
		return nil, nil
	}

	lex := NewLex(src, p.Config)

	var opts *Options
	if inst != nil {
		opts = inst.options
	}

	ctx := &Context{
		UI:       0,
		T0:       NoToken,
		T1:       NoToken,
		V1:       NoToken,
		V2:       NoToken,
		RS:       make([]*Rule, len(src)*4+100),
		RSI:      0,
		RSM:      p.RSM,
		Meta:     meta,
		LexSubs:  lexSubs,
		RuleSubs: ruleSubs,
		Opts:     opts,
		Cfg:      p.Config,
		Src:      src,
		Inst:     inst,
		U:        make(map[string]any),
		TC:       0,
		NOTOKEN:  NoToken,
		NORULE:   NoRule,
		F:        func(v any) string { return Str(v, 44) },
	}

	lex.Ctx = ctx

	startName := p.Config.RuleStart
	if startName == "" {
		startName = "val"
	}
	startSpec := p.RSM[startName]
	if startSpec == nil {
		return nil, nil
	}

	rule := MakeRule(startSpec, ctx, nil)
	root := rule
	ctx.Root = root

	// Run parse.prepare hooks
	if len(p.Config.ParsePrepare) > 0 {
		for _, prep := range p.Config.ParsePrepare {
			prep(ctx)
		}
	}

	// Maximum iterations: 2 * numRules * srcLen * 2 * maxmul
	maxmul := p.MaxMul
	if maxmul <= 0 {
		maxmul = 3
	}
	maxr := 2 * len(p.RSM) * len(src) * 2 * maxmul
	if maxr < 100 {
		maxr = 100
	}

	kI := 0
	for rule != NoRule && kI < maxr {
		ctx.KI = kI
		ctx.Rule = rule

		// Fire rule subscribers BEFORE process (matching TS).
		if len(ctx.RuleSubs) > 0 {
			for _, sub := range ctx.RuleSubs {
				sub(rule, ctx)
			}
		}

		rule = rule.Process(ctx, lex)

		// Check for parse error from alt.E or actions.
		if ctx.ParseErr != nil {
			// Prefer lexer errors (e.g. unterminated_string) over generic
			// "unexpected" from alt matching, since the lex error is more
			// specific about what went wrong. Matches TS behavior where
			// lex errors propagate through #ZZ tokens.
			if lex.Err != nil {
				return nil, lex.Err
			}
			tkn := ctx.ParseErr
			return nil, p.makeError("unexpected", tkn.Src, src, tkn.SI, tkn.RI, tkn.CI)
		}

		kI++
	}

	// Check for lexer errors (unterminated strings, comments, etc.)
	if lex.Err != nil {
		return nil, lex.Err
	}

	// Check for unconsumed tokens (syntax error) - explicit trailing content check.
	// First check tokens already in the lookahead buffer.
	if ctx.T0 != nil && !ctx.T0.IsNoToken() && ctx.T0.Tin != TinZZ {
		return nil, p.makeError("unexpected", ctx.T0.Src, src, ctx.T0.SI, ctx.T0.RI, ctx.T0.CI)
	}
	// Also explicitly ask lexer for more (matching TS parser.ts:187-189).
	endTkn := lex.Next(rule)
	if endTkn.Tin != TinZZ {
		return nil, p.makeError("unexpected", endTkn.Src, src, endTkn.SI, endTkn.RI, endTkn.CI)
	}
	// Check lexer errors from that final Next() call.
	if lex.Err != nil {
		return nil, lex.Err
	}

	// Follow replacement chain: when val is replaced by list (implicit list),
	// root.Node is stale. Follow Next/Prev links to find the actual result.
	result := root
	for result.Next != NoRule && result.Next != nil && result.Next.Prev == result {
		result = result.Next
	}

	if IsUndefined(result.Node) {
		return nil, nil
	}

	// Check result.fail
	if len(p.Config.ResultFail) > 0 {
		for _, fail := range p.Config.ResultFail {
			if result.Node == fail {
				return nil, p.makeError("unexpected", "", src, 0, 1, 1)
			}
		}
	}

	return result.Node, nil
}

// makeError creates a JsonicError using this parser's error messages.
func (p *Parser) makeError(code, src, fullSource string, pos, row, col int) *JsonicError {
	msgs := p.ErrorMessages
	if msgs == nil {
		msgs = errorMessages
	}
	tmpl, ok := msgs[code]
	if !ok {
		tmpl = msgs["unknown"]
		if tmpl == "" {
			tmpl = errorMessages["unknown"]
		}
	}
	detail := tmpl + src

	hint := ""
	if p.Hints != nil {
		hint = p.Hints[code]
	}

	return &JsonicError{
		Code:       code,
		Detail:     detail,
		Pos:        pos,
		Row:        row,
		Col:        col,
		Src:        src,
		Hint:       hint,
		fullSource: fullSource,
		tag:        p.ErrTag,
	}
}

// parseNumericString converts a numeric string to float64.
// Handles standard decimals, hex (0x), octal (0o), binary (0b), and signs.
func parseNumericString(s string) float64 {
	if len(s) == 0 {
		return math.NaN()
	}

	// Handle sign prefix for special formats
	sign := 1.0
	ns := s
	if ns[0] == '-' {
		sign = -1.0
		ns = ns[1:]
	} else if ns[0] == '+' {
		sign = 1.0
		ns = ns[1:]
	}

	if len(ns) >= 2 {
		switch {
		case ns[0] == '0' && (ns[1] == 'x' || ns[1] == 'X'):
			val, err := strconv.ParseInt(ns[2:], 16, 64)
			if err != nil {
				return math.NaN()
			}
			return sign * float64(val)
		case ns[0] == '0' && (ns[1] == 'o' || ns[1] == 'O'):
			val, err := strconv.ParseInt(ns[2:], 8, 64)
			if err != nil {
				return math.NaN()
			}
			return sign * float64(val)
		case ns[0] == '0' && (ns[1] == 'b' || ns[1] == 'B'):
			val, err := strconv.ParseInt(ns[2:], 2, 64)
			if err != nil {
				return math.NaN()
			}
			return sign * float64(val)
		}
	}

	// Remove underscores if present
	ns = strings.ReplaceAll(s, "_", "")

	val, err := strconv.ParseFloat(ns, 64)
	if err != nil {
		return math.NaN()
	}

	// Normalize -0 to 0
	if val == 0 {
		return 0
	}

	return val
}
